src/webapp/components/dataset-wizard/DisaggregationStep.tsx

```diff
@@ -0,0 +1,200 @@
+
+                const currentFullId = [
+                    indicator.id,
+                    ...dataElements.map(dataElement => dataElement.id),
+                ].join(".");
+                const selectedFullId = [
+                    selectedIndicator.indicator.id,
+                    ...selectedIndicator.dataElements.map(dataElement => dataElement.id),
+                ].join(".");
+
+                const newIndicator = Indicator.create({
+                    ...indicator,
+                    categories: selectedCategories,
+                    relatedDataElements:
+                        selectedIndicator.indicator.type === "outcomes"
+                            ? dataElements.map(dataElement => {
+                                  return { ...dataElement, categories: selectedCategories };
+                              })
+                            : [],
+                });
+                const newDataElements = dataElements.map(dataElement => {
+                    return { ...dataElement, categories: selectedCategories };
+                });
+
+                const record: IndicatorWithDataElement = {
+                    indicator: newIndicator,
+                    dataElements: newDataElements,
+                };
+
+                if (mode === "indicator" && currentFullId === selectedFullId) {
+                    return record;
+                } else if (
+                    mode === "competency" &&
+                    indicator.coreCompetency.id === selectedIndicator.indicator.coreCompetency.id
+                ) {
+                    return record;
+                } else if (mode === "all") {
+                    return record;
+                } else {
+                    return { indicator, dataElements };
+                }
+            }
+        );
+
+        setIndicatorsDataElements(updatedIndicator);
+        setSelectedIndicator(undefined);
+        onChange(dataSet.setIndicators(updateIndicatorsDataElements(dataSet, updatedIndicator)));
+    };
+
+    return (
+        <div>
+            <SearchBox value={search} onChange={setSearch} />
+
+            {filteredIndicators.map(indicatorWithDataElement => {
+                const { indicator, dataElements } = indicatorWithDataElement;
+
+                const ids = dataElements.map(dataElement => dataElement.id).join(".");
+
+                const disaggregationName = _(dataElements)
+                    .map(dataElement => dataElement.disaggregation?.name)
+                    .uniq()
+                    .join("/");
+
+                const categoriesNames = dataElements
+                    ? _(dataElements)
+                          .map(dataElement => {
+                              return dataElement.categories.map(category => category.name);
+                          })
+                          .flatten()
+                          .uniq()
+                          .join("/")
+                    : "";
+
+                return (
```

Comment:
too much code to be in JSX, that's a signal we need to abstract the component in the block.